docs/bundle_format.rst

```diff
@@ -0,0 +1,106 @@
+Bundle Format
+================
+
+The Blockstore does not parse or validate most files itself. However, some conventions must be followed by clients
+to allow interoperability between environments:
+
+* There are no constraints on how the files in a Bundle can be organized.
+* OLX files must use the **.olx** extension. Files with a different extension may be ignored or not treated as
+  OLX by plugins and clients.
+* A bundle.json file must be added by authoring clients which specifies the shareable components, bundle
+  dependencies and assets. See below for details.
+* Every XBlock element in an OLX file must have an **id** attribute. The value of this attribute will be used
+  to construct the **block_key**. In a single Bundle, all ids must be unique. Blockstore will not validate ids however.
```

Comment:
Basically decoupling filenames and ids. Though would it make sense to keep them coupled?

Reply:
Checking that IDs in a bundle are unique might be slightly expensive, if it requires scanning the contents of all OLX files in the bundle.

It would be much easier to just say that the runtime will generate the full unique ID for each XBlock from the OLX file name combined with an optional id attribute specified within the OLX. But that would not be backwards compatible with current courses, which allow authors to specify any custom ID slug (`url_name`) for each XBlock.

Although maybe we can collect+index+check IDs during the compilation phase (["Transform 1" in my proposal](https://openedx.atlassian.net/wiki/spaces/AC/pages/803405867/Compositor+Architecture+Proposal)). That would be efficient.

Reply:
Hmm. Yeah, this is a bit tricky. To be clear my since Blockstore is not expected to parse the files so it will not be verifying them. Anything inside the OLX file will have to be verified at some other step.

By file name do you mean just the filename or the whole path because if the former then we will need to make sure it is unique across all directories. And whichever it is, changing it implies deleting the block and creating a new one. Right?

Reply:
> By file name do you mean just the filename or the whole path

I guess I mean the whole path, slugified. And yeah, if a block ID changes, it's effectives the same as deleting it and creating a new one as far as the LMS is concerned.

Reply:
@ormsbee Did you look at this id generation idea? Just want to make sure we are all on the same page.